core/commands/p2p.go

```diff
@@ -296,31 +313,36 @@ var p2pCloseCmd = &cmds.Command{
 		}
 
 		match := func(listener p2p.Listener) bool {
-			out := true
-
-			if p {
-				out = out && (proto == listener.Protocol())
+			if closeAll {
+				return true
 			}
-			if l {
-				out = out && (listen == listener.ListenAddress())
+			if p && proto != listener.Protocol() {
+				return false
 			}
-			if t {
-				out = out && (target == listener.TargetAddress())
+			if l && !listen.Equal(listener.ListenAddress()) {
+				return false
 			}
-
-			out = out || closeAll
-			return out
+			if t && !target.Equal(listener.TargetAddress()) {
+				return false
+			}
+			return true
 		}
 
-		var closed int
-		for _, listener := range n.P2P.Listeners.Listeners {
-			if !match(listener) {
+		todo := make([]p2p.Listener, 0)
+		n.P2P.Listeners.Lock()
+		for _, l := range n.P2P.Listeners.Listeners {
+			if !match(l) {
 				continue
 			}
-			listener.Close()
-			closed++
+			todo = append(todo, l)
 		}
-		res.SetOutput(closed)
+		n.P2P.Listeners.Unlock()
+
+		for _, l := range n.P2P.Listeners.Listeners {
```

Comment:
Range over todo.

Reply:
We should probably also have a test case (sharness?) that checks to make sure we don't close the wrong listeners.